enhancements/machine-api/remediation-history.md

```diff
@@ -0,0 +1,168 @@
+
+## Summary
+
+Record and show limited remediation history.
+
+## Motivation
+
+With machine health checks administrators can configure automatic remediation of node problems. There is no good
+overview though for when and why nodes were remediated. Recording and showing a history of remediations would allow
+administrators to review when and why machines were remediated, and to discover patterns like e.g. machineA always
+reboots at 2am on Wednesdays.
+
+### Goals
+
+- Record a limited remediation history.
+
+- Show remediation history in the UI.
+
+### Non-Goals
+
+- Record an unlimited history of remediations.
+
+## Proposal
+
+The machine health check holds the configuration for automatic remediation of nodes. We want to extend the status of
+machine health checks to contain the last x remediation events, with information about which node is affected, which
+condition triggered the remediation, and the timestamps of
+- when the triggering condition was detected
+- when remediation was started
+- when the node was succesfully fenced (deleted / powered off)
+- when the node is healthy again.
+
+This information can be displayed in the UI in a table on a machine healthcheck details page.
+
+Example MHC:
+
+```yaml
+apiVersion: machine.openshift.io/v1beta1
+kind: MachineHealthCheck
+metadata:
+  annotations:
+    machine.openshift.io/remediation-strategy: external-baremetal
+  creationTimestamp: "2020-11-24T16:31:22Z"
+spec:
+  maxUnhealthy: 100%
+  nodeStartupTimeout: 60m
+  selector:
+    matchLabels:
+      machine.openshift.io/cluster-api-machine-role: worker
+  unhealthyConditions:
+  - status: "False"
+    timeout: 20s
+    type: Ready
+  - status: Unknown
+    timeout: 20s
+    type: Ready
+status:
+  conditions:
+  - lastTransitionTime: "2020-11-24T17:17:32Z"
+    status: "True"
+    type: RemediationAllowed
+  currentHealthy: 2
+  expectedMachines: 2
+  remediationHistory:
+  - conditionStatus: Unknown
+    conditionType: Ready
+    detected: "2020-11-24T17:33:43Z"
+    fenced: "2020-11-24T17:34:35Z"
+    finished: "2020-11-24T17:35:09Z"
+    remediationType: external
+    started: "2020-11-24T17:34:04Z"
+    targetKind: Node
+    targetName: worker-1
+  remediationsAllowed: 2
+```
+
+Example UI: TBD (copy from "KNI Node & Host Management Designs" doc?)
+
+### User Stories
+
+#### Story 1
+
+As a cluster administrator I want to have an overview over when and why nodes failed and were remediated, in order
+to detect patterns, and to be able to troubleshoot possible root causes.
+
+### Implementation Details/Notes/Constraints [optional]
+
+The machine healthcheck controller [0] is responsible for triggering remediations based on machine health checks and node
+conditions. This will also be the place to track remediations:
+
+- in the `needsRemediation` func:
+  - for failed machines and unhealthy node conditions: track detection
+  - for missing nodes after remdiation start: track successful fencing
```

Comment:
What happens, if for example, fencing happens correctly, but there is an api conflict and as such, the status update is not persisted, can this be re-calculated on the next reconciliation loop?

Reply:
Good question. I think it might be hard to catch all intermediate states in this case (e.g. we record that a condition is unhealthy, but fail to update the remediation start). However, we should always be able to do a "cleanup" when the node is back to healthy,  by triggering reconcile until the status update succeeds. Good enough?

Reply:
I guess the question is, is it bad if the remediation gets restarted here? If there's no prior record and the MHC sees the Node is still unhealthy, will it re-fence/remediate it?

Reply:
The intention of the remediation history is just to track what's happening for documentation purposes. It should not have any effect on deciding what should happen next, that will continue to be only based on the actual cluster state.
Does that answer the question?

Reply:
Essentially, there is no plan to use these status entries to make decisions in subsequent reconcile loops.
Worst case, it might look like the node came back healthy on its own, or some external entity deleted the Machine for us.